### ISSUE TITLE
Linting a Vue SFC containing JSX throws a parsing error

### ISSUE THREAD

**Issue by leonzalion** [vuejs/vue-eslint-parser#150]
The following Vue SFC causes a lint error:

```vue
<script setup lang="jsx">
const JsxElement = (
	<div>
		Rendered from JSX!
	</div>
)
</script>

<template>
	<JsxElement />
</template>
```

![image](https://user-images.githubusercontent.com/36966635/163624809-1adc3a92-6f4e-456f-8e59-f01e319ced5e.png)

**Repro:** https://github.com/leonzalion/vue-eslint-plugin-jsx-bug

I did some debugging, and I think the problem has to do with the .vue file extension when the file is passed to `@typescript-eslint/parser`.

This is the `parserOptions` configuration passed to `@typescript-eslint/parser` at this line: https://github.com/vuejs/vue-eslint-parser/blob/c476e7c25ab161b336339448765085e6219d0e04/src/script/index.ts#L555

```
 {
  comment: true,
  loc: true,
  range: true,
  tokens: true,
  ecmaVersion: undefined,
  sourceType: 'module',
  parser: '@typescript-eslint/parser',
  project: '/Users/leonzalion/projects/jsx-test/tsconfig.json',
  extraFileExtensions: [ '.vue' ],
  ecmaFeatures: { jsx: true, globalReturn: false },
  raw: true,
  eslintVisitorKeys: true,
  eslintScopeManager: true,
  filePath: '/Users/leonzalion/projects/jsx-test/src/app.vue'
}
```

When I modified the JavaScript code and only changed the above parserOptions.filePath parameter from `app.vue` to `app.jsx`, the linter didn't throw any error.

I then traced down the bug through the `@typescript-eslint/parser` code and I think it has to do with the following line: https://github.com/typescript-eslint/typescript-eslint/blob/5e794512bf124c39de76d4e2cf8a3d6cfb08f1a8/packages/typescript-estree/src/create-program/createWatchProgram.ts#L351

I think this line causes TypeScript to try and infer the type of the file based on the file extension, and because the file extension in `.vue` (which TypeScript doesn't recognize), TypeScript treats the file as a `.ts` file by default, causing the parsing to fail since the file is actually a JSX file.

(Sidenote: when I tried changing that line to `ts.ScriptKind.TSX`, it gave me the following error:)
![Screen Shot 2022-04-15 at 3 58 37 PM](https://user-images.githubusercontent.com/36966635/163626432-c03779b7-52c3-4bcb-b729-4add470035c3.png)

When I modified the TypeScript compiler code and added a `case ".vue":` under [this line](https://github.com/microsoft/TypeScript/blob/0d9a4fe490499de6ec4b8a399ec96645b67d99a6/src/compiler/utilities.ts#L6957), the lint then started working again (you can test this by running `npx patch-package && npm lint` in the [repro](https://github.com/leonzalion/vue-eslint-plugin-jsx-bug)).

As somebody who isn't really experienced with the internals of ESLint, I'm not exactly sure what the best way to fix this is, so please let me know what you think!

**Comment by douira** [comment]
I tried `jsx: false` as suggested in this related issue https://github.com/vuejs/eslint-plugin-vue/issues/1223 but it didn't fix it for me. I think I'm experiencing the same issue as you are.

**Comment by ota-meshi** [comment]
Thank you for posting this issue.

I think it's a `typescript-eslint/parser` spec. It is documented in the following link.
https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/parser#parseroptionsecmafeaturesjsx

I think you need to stop using JSX in `.vue` file, empty `parserOptions.project`, or use another parser.

**Comment by ota-meshi** [comment]
By the way, even if we change the extension virtually on the parser, parsing will fail because it is a file that does not actually exist.

**Comment by leonzalion** [comment]
Would it be possible to do something like `vue-tsc` where they "patch" the list of extensions? https://github.com/johnsoncodehk/volar/blob/master/packages/vue-tsc/bin/vue-tsc.js

**Comment by ota-meshi** [comment]
I'm not familiar with the API inside typescript so I don't know :sweat:

**Comment by leonzalion** [comment]
> When I modified the TypeScript compiler code and added a case ".vue": under [this line](https://github.com/microsoft/TypeScript/blob/0d9a4fe490499de6ec4b8a399ec96645b67d99a6/src/compiler/utilities.ts#L6957), the lint then started working again (you can test this by running npx patch-package && npm lint in the [repro](https://github.com/leonzalion/vue-eslint-plugin-jsx-bug)).

I'm not sure if this is a foolproof solution, but maybe it might be possible to do that at runtime? Would a PR attempting to do that be accepted?

**Comment by ota-meshi** [comment]
I don't accept the patch approach PR because I don't think it's guaranteed. I think you'll need to ask the TypeScript team to add a guaranteed API if needed.

**Comment by ota-meshi** [comment]
You may be able to patch it personally by putting the following script in `.eslintrc.js`.

```js
        const ts = require('typescript')
        const { ensureScriptKind } = ts
        ts.ensureScriptKind = function (fileName, ...args) {
            if (fileName.endsWith(".vue")) {
                return ts.ScriptKind.TSX
            }
            return ensureScriptKind.call(this, fileName, ...args)
        }
```

**Comment by brilik** [comment]
You can try uses default settings `.eslintrc.js` after install step-by-step vue.js with `npm`.
```
// .eslintrc.js

'extends': [
    'plugin:vue/vue3-essential',
    'eslint:recommended',
    '@vue/eslint-config-typescript',
    '@vue/eslint-config-prettier/skip-formatting'
  ],
  parserOptions: {
    ecmaVersion: 'latest'
  },
```

**Comment by eric-gitta-moore** [comment]
same here.

link https://github.com/vuejs/vue-eslint-parser/issues/45#issuecomment-3130626719
link https://github.com/vuejs/eslint-plugin-vue/issues/2304#issuecomment-3130586938
